### PR TITLE
output: Don't serialize `hash_info` or `meta` in cloud versioning.

### LINKED ISSUE #8357: cloud versioning: get rid of `.dir` files
Part of #7995

With cloud versioned remotes, the resulting `.dvc` file for directories looks like:

```yaml
outs:
- md5: f437247ec66d73ba66b0ade0246fcb49.dir
  size: 0
  nfiles: 2
  path: data
  files:
  - size: 2
    version_id: mHkZWv4Pa0rZm0SaFdew0WKu7cxPUHrW
    etag: b026324c6904b2a9cb4b88d6d61c81d1
    md5: b026324c6904b2a9cb4b88d6d61c81d1
    relpath: '1'
  - size: 2
    version_id: GO04I1aowL.OoiDYYcoX90W96PufrgFE
    etag: 26ab0db90d72e28ad0ba1e22ee510510
    md5: 26ab0db90d72e28ad0ba1e22ee510510
    relpath: '2'
```

There is a corresponding entry in `.dvc/cache/f4/37247ec66d73ba66b0ade0246fcb49.dir`:

```json
[{"md5": "b026324c6904b2a9cb4b88d6d61c81d1", "relpath": "1"}, {"md5": "26ab0db90d72e28ad0ba1e22ee510510", "relpath": "2"}]
```

Is the `.dir` entry needed?

### PR DESCRIPTION
**Pull request by daavoo**
Closes #8357

---

Here is how the `.dvc` file looks for `2.42.0` / `this P.R` at different steps of the workflow (click details to expand):

## Importing with `--version-aware`

```
$ dvc import-url --version-aware $VERSIONED_BUCKET/data
$ cat data.dvc
```

<details>
<table>
<tr>
<td> 2.42.0 </td> <td> #8849 </td>
</tr>
<tr>
<td>

```yaml
md5: 677db99135cf82ed1b726b6be7abef49
frozen: true
deps:
- md5: a042035f7fb17fc84a42e33787ae4b6a.dir
  size: 13
  nfiles: 3
  path: s3://diglesia-bucket/data
  files:
  - size: 5
    version_id: dQymcW0EvEqy23qfrksrqo0ni2zGtPMG
    etag: 042967ff088f7f436015fbebdcad1f28
    relpath: bar
  - size: 4
    version_id: pd67.5YcvjqDm7jfmiHbpeQsXnprrXtf
    etag: d3b07384d113edec49eaa6238ad5ff00
    relpath: foo
  - size: 4
    version_id: 2ruNQaorqjnsFQZ5cinHsZSHEw7cz26J
    etag: 7361528e901ca2f2f1952a68ad242d79
    relpath: goo
outs:
- md5: b5c587798f2c0f599ef935244b290bf1.dir
  size: 13
  nfiles: 3
  path: data
  push: false
```

</td>
<td>

```yaml
md5: f642080c781ca41a33a3e6bf77959af5
frozen: true
deps:
- files:
  - size: 5
    version_id: dQymcW0EvEqy23qfrksrqo0ni2zGtPMG
    etag: 042967ff088f7f436015fbebdcad1f28
    relpath: bar
  - size: 4
    version_id: pd67.5YcvjqDm7jfmiHbpeQsXnprrXtf
    etag: d3b07384d113edec49eaa6238ad5ff00
    relpath: foo
  - size: 4
    version_id: 2ruNQaorqjnsFQZ5cinHsZSHEw7cz26J
    etag: 7361528e901ca2f2f1952a68ad242d79
    relpath: goo
  path: s3://diglesia-bucket/data
outs:
- md5: b5c587798f2c0f599ef935244b290bf1.dir
  size: 13
  nfiles: 3
  path: data
  push: false
```

</td>
</tr>
</table>
</details>

## Tracking local folder with `dvc add` and running the first `dvc push`

```
$ mkdir data
$ echo bar > data/bar
$ echo foo > data/foo
$ dvc add data
$ dvc push -v
$ cat data.dvc
```

<details>
<table>
<tr>
<td> 2.42.0 </td> <td> #8849 </td>
</tr>
<tr>
<td>

```yaml
outs:
- md5: 168fd6761b9c3f2c081085b51fd7bf15.dir
  size: 8
  nfiles: 2
  path: data
  files:
  - size: 4
    version_id: iGoUk9Wlb95nNOw6iPHUrnNWq9E898YO
    etag: c157a79031e1c40f85931829bc5fc552
    md5: c157a79031e1c40f85931829bc5fc552
    relpath: bar
  - size: 4
    version_id: DwITkoYE8ipINKzK2QcGKvlwlqaS4oER
    etag: d3b07384d113edec49eaa6238ad5ff00
    md5: d3b07384d113edec49eaa6238ad5ff00
    relpath: foo
```

</td>
<td>

```yaml
outs:
- path: data
  files:
  - size: 4
    version_id: DhE9omrAq7IXC5wGIIGqruq9JyP_Nwhm
    etag: c157a79031e1c40f85931829bc5fc552
    md5: c157a79031e1c40f85931829bc5fc552
    relpath: bar
  - size: 4
    version_id: Vnu6l9_ECdLGrp6B5iK6YmQTm87UaDzP
    etag: d3b07384d113edec49eaa6238ad5ff00
    md5: d3b07384d113edec49eaa6238ad5ff00
    relpath: foo
```

</td>
</tr>
</table>
</details>

## Tracking a local update with `dvc add` on that folder

```
$ echo bar2 > data/bar
$ dvc add data
$ cat data.dvc
```

<details>
<table>
<tr>
<td> 2.42.0 </td> <td> #8849 </td>
</tr>
<tr>
<td>

```yaml
outs:
- md5: 43f06a6517d71d0267bed23cfb1a53fb.dir
  size: 9
  nfiles: 2
  path: data
  files:
  - size: 5
    md5: 042967ff088f7f436015fbebdcad1f28
    relpath: bar
  - size: 4
    version_id: DwITkoYE8ipINKzK2QcGKvlwlqaS4oER
    etag: d3b07384d113edec49eaa6238ad5ff00
    md5: d3b07384d113edec49eaa6238ad5ff00
    relpath: foo
```

</td>
<td>

```yaml
outs:
- path: data
  files:
  - size: 5
    md5: 042967ff088f7f436015fbebdcad1f28
    relpath: bar
  - size: 4
    version_id: Vnu6l9_ECdLGrp6B5iK6YmQTm87UaDzP
    etag: d3b07384d113edec49eaa6238ad5ff00
    md5: d3b07384d113edec49eaa6238ad5ff00
    relpath: foo
```

</td>
</tr>
</table>
</details>

## After pushing the local update with `dvc push`

```
$ dvc push 
$ cat data.dvc
```

<details>
<table>
<tr>
<td> 2.42.0 </td> <td> #8849 </td>
</tr>
<tr>
<td>

```yaml
outs:
- md5: 43f06a6517d71d0267bed23cfb1a53fb.dir
  size: 9
  nfiles: 2
  path: data
  files:
  - size: 5
    md5: 042967ff088f7f436015fbebdcad1f28
    relpath: bar
    version_id: QLNufQhPEwlSlNSfauemR5D23RRMSbfF
    etag: 042967ff088f7f436015fbebdcad1f28
  - size: 4
    version_id: DwITkoYE8ipINKzK2QcGKvlwlqaS4oER
    etag: d3b07384d113edec49eaa6238ad5ff00
    md5: d3b07384d113edec49eaa6238ad5ff00
    relpath: foo
```

</td>
<td>

```yaml
outs:
- path: data
  files:
  - size: 5
    md5: 042967ff088f7f436015fbebdcad1f28
    relpath: bar
  - size: 4
    version_id: Vnu6l9_ECdLGrp6B5iK6YmQTm87UaDzP
    etag: d3b07384d113edec49eaa6238ad5ff00
    md5: d3b07384d113edec49eaa6238ad5ff00
    relpath: foo
```

</td>
</tr>
</table>
</details>

## Tracking a modification in the remote with `dvc update`

```
$ echo goo > goo
$ aws s3 cp goo $REMOTE/data/goo
$ dvc update data.dvc -v
$ cat data.dvc
```

<details>
<table>
<tr>
<td> 2.42.0 </td> <td> #8849 </td>
</tr>
<tr>
<td>

```yaml
outs:
- md5: b5c587798f2c0f599ef935244b290bf1.dir
  size: 13
  nfiles: 3
  path: data
  files:
  - size: 5
    version_id: QLNufQhPEwlSlNSfauemR5D23RRMSbfF
    etag: 042967ff088f7f436015fbebdcad1f28
    md5: 042967ff088f7f436015fbebdcad1f28
    relpath: bar
  - size: 4
    version_id: DwITkoYE8ipINKzK2QcGKvlwlqaS4oER
    etag: d3b07384d113edec49eaa6238ad5ff00
    md5: d3b07384d113edec49eaa6238ad5ff00
    relpath: foo
  - size: 4
    version_id: xktZ1CcLphMaZheT_O9QrxFpmd9GrgK1
    etag: 7361528e901ca2f2f1952a68ad242d79
    md5: 7361528e901ca2f2f1952a68ad242d79
    relpath: goo
```

</td>
<td>

```yaml
outs:
- path: data
  files:
  - size: 5
    version_id: dQymcW0EvEqy23qfrksrqo0ni2zGtPMG
    etag: 042967ff088f7f436015fbebdcad1f28
    md5: 042967ff088f7f436015fbebdcad1f28
    relpath: bar
  - size: 4
    version_id: pd67.5YcvjqDm7jfmiHbpeQsXnprrXtf
    etag: d3b07384d113edec49eaa6238ad5ff00
    md5: d3b07384d113edec49eaa6238ad5ff00
    relpath: foo
  - size: 4
    version_id: 2ruNQaorqjnsFQZ5cinHsZSHEw7cz26J
    etag: 7361528e901ca2f2f1952a68ad242d79
    md5: 7361528e901ca2f2f1952a68ad242d79
    relpath: goo
```

</td>
</tr>
</table>
</details>

### CODE REVIEW

File: dvc/output.py [hunk 1]
```diff
@@ -368,7 +368,11 @@ def __init__(
             name=self.hash_name,
             value=getattr(self.meta, self.hash_name, None),
         )
-        if self.meta.nfiles or self.hash_info and self.hash_info.isdir:
+        if self.files:
+            tree = Tree.from_list(self.files, hash_name=self.hash_name)
+            tree.digest()
+            self.hash_info = tree.hash_info
```
Comment (daavoo) [inline dvc/output.py:374]: This introduces overhead but need to check how big it is

Comment (daavoo) [reply dvc/output.py:374]: Also, `self.meta` is being lost here, before it was restored from the `.dvc` file. 
Is still unclear to me how `meta` is used and if it's needed at all in this scenario

**Review by dberenbaum — APPROVED** [review]

File: dvc/output.py [hunk 2]
```diff
@@ -747,7 +754,22 @@ def _commit_granular_dir(self, filter_info) -> Optional["HashFile"]:
     def dumpd(self, **kwargs):  # noqa: C901
         meta = self.meta.to_dict()
         meta.pop("isdir", None)
-        ret: Dict[str, Any] = {**self.hash_info.to_dict(), **meta}
+        ret: Dict[str, Any] = {}
+        if (
+            (not self.IS_DEPENDENCY or self.stage.is_import)
+            and self.hash_info.isdir
+            and (kwargs.get("with_files") or self.files is not None)
+        ):
```
Comment (skshetry) [inline dvc/output.py:762]: Any reason to move this upward @daavoo? This might affect the ordering of metadata in `.dvc` file.

Comment (daavoo) [reply dvc/output.py:762]: No real reason. Could be moved bellow

Comment (skshetry) [reply dvc/output.py:762]: Created https://github.com/iterative/dvc/pull/8879.

**Pull request merged**